server.js

```diff
@@ -730,167 +729,131 @@ app.post('/submit-match',  asyncMiddleware( async (req, res, next) => {
+                        ip: req.ip, winnerhash: req.body.winnerhash, loserhash: req.body.loserhash, sgf: sgfbuffer.toString(),
+                        options_hash: req.body.options_hash,
+                        clientversion: Number(req.body.clientversion), winnercolor: req.body.winnercolor,
+                        movescount: (req.body.movescount ? Number(req.body.movescount) : null),
+                        score: req.body.score,
+                        random_seed: req.body.random_seed
                     }
-                }
+                },
+                { upsert: true }
+            );
+
+            // Not inserted, we got duplicate sgfhash, abort!
+            if (!dbres.upsertedId) {
+                return reject('Upload match with duplicate sgf');
             }
-        );
-      }
+
+            console.log(req.ip + " (" + req.headers['x-real-ip'] + ") " + " uploaded match " + sgfhash);
+            res.send("Match data " + sgfhash + " stored in database\n");
+            resolve();
+        });
     });
 
+    // prepare $inc 
+    var $inc = { game_count: 1 };
+    if (match.network1 == req.body.winnerhash)
+        $inc.network1_wins = 1;
+    else
+        $inc.network1_losses = 1;
+
+    // save to database using $inc and get modified document
+    match = (await db.collection("matches").findOneAndUpdate(
+        { _id: match._id },
+        { $inc: $inc },
+        { returnOriginal: false }  // return modified document 
+    )).value;
+
+    // get latest SPRT result
+    const sprt_result = SPRT(match.network1_wins, match.network1_losses);
+    var pending_match_index = pending_matches.findIndex(m => m._id.equals(match._id));
+
+    // match is found in pending_matches
+    if (pending_match_index >= 0) {
+        if (sprt_result === false) {
+            // remove from pending matches
+            console.log("SPRT: Early fail pop: " + JSON.stringify(pending_matches.splice(pending_match_index, 1)));
+            console.log("SPRT: Early fail post-pop: " + JSON.stringify(pending_matches));
+        } else {
+            var m = pending_matches[pending_match_index];
+
+            // remove the match from the requests array.
+            var index = m.requests.findIndex(e => e.seed === seed_from_mongolong(req.body.random_seed));
+            if (index !== -1) {
+                m.requests.splice(index, 1);
+            }
+
+            // update stats 
+            m.game_count++;
+            if (m.network1 == req.body.winnerhash) {
+                m.network1_wins++;
+            } else {
+                m.network1_losses++;
+            }
+
+            // Check > 1 since we'll run to 400 even on a SPRT pass, but will do it at end.
+            //
+            if (sprt_result === true && pending_matches.length > 1) {
+                console.log("SPRT: Early pass unshift: " + JSON.stringify(m));
+                console.log("SPRT: Early pass post-unshift: " + JSON.stringify(pending_matches));
```

Comment:
There used to be a `pending_matches.unshift(pending_matches.splice(pending_match_index, 1));` in the first commit that got removed in https://github.com/gcp/leela-zero-server/pull/64/commits/4d57ea6a3cd9452157c93eb4714efb1dbe854e37. Shouldn't that be here to keep the passing match for later?